File: app/src/main/java/com/digicoffer/lauditor/LoginActivity/ViewModels/reset_password_file.java
package com.digicoffer.lauditor.LoginActivity.ViewModels;

import static com.digicoffer.lauditor.CommonFiles.GlobalFiles.Constants.check_url;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.core.view.WindowCompat;

import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.ColorStateList;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.digicoffer.lauditor.R;
import com.digicoffer.lauditor.Webservice.AsyncTaskCompleteListener;
import com.digicoffer.lauditor.Webservice.HttpResultDo;
import com.digicoffer.lauditor.Webservice.CommonApiHelper.WebServiceHelper;
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils;
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.Constants;
import com.google.android.material.textfield.TextInputEditText;

import org.json.JSONObject;

import java.util.Objects;

public class reset_password_file extends AppCompatActivity implements AsyncTaskCompleteListener {

    private Dialog progressDialog;
    private TextInputEditText password1, password2;
    private Button submit;
    private TextView tvErrorMsg;
    ImageView iv_toggle_password1, iv_toggle_password2;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.reset_password_file);

        getWindow().setStatusBarColor(ContextCompat.getColor(this, R.color.Blue_text_color));
        WindowCompat.getInsetsController(getWindow(), getWindow().getDecorView())
                .setAppearanceLightStatusBars(true);
        if (getSupportActionBar() != null) {
            getSupportActionBar().hide();
        }

        // Check if this is a forced password reset
        boolean isResetMode = getIntent().getBooleanExtra("reset_mode", false);

        tvErrorMsg       = findViewById(R.id.tv_error_msg);
        password1        = findViewById(R.id.et_login_password);
        iv_toggle_password1 = findViewById(R.id.iv_toggle_password);

        LinearLayout ll_password2 = findViewById(R.id.ll_password2);
        password2        = ll_password2.findViewById(R.id.et_login_password);
        iv_toggle_password2 = ll_password2.findViewById(R.id.iv_toggle_password);

        password1.setHint(R.string.password);
        password2.setHint(R.string.confirm_password);

        submit = findViewById(R.id.Submit);
        submit.setEnabled(false);
        submit.setText(R.string.reset);
        submit.setBackgroundTintList(ColorStateList.valueOf(
                getResources().getColor(R.color.dullBlueColor)));

        // Load saved data if in reset mode
        if (isResetMode) {
            SharedPreferences prefs = getSharedPreferences("MyPrefs", Context.MODE_PRIVATE);
            Constants.PK      = prefs.getString("pk", "");
            Constants.USER_ID = prefs.getString("user_id", "");
            Constants.TOKEN   = prefs.getString("Token", "");
        }

        // Show all 3 conditions as soon as password1 is focused
        tvErrorMsg.setVisibility(View.VISIBLE);
        tvErrorMsg.setText(
                "• Must be 8–15 characters long\n" +
                        "• Must include uppercase and lowercase letters\n" +
                        "• Must include a number and a special character"
        );
//        password1.setOnFocusChangeListener((v, hasFocus) -> {
//            if (hasFocus) {
//                tvErrorMsg.setVisibility(View.VISIBLE);
//                tvErrorMsg.setText(
//                        "• Must be 8–15 characters long\n" +
//                                "• Must include uppercase and lowercase letters\n" +
//                                "• Must include a number and a special character"
//                );
//            }
//        });

        // password1 watcher — validates conditions one by one
        password1.addTextChangedListener(new TextWatcher() {
            @Override public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                checkPasswordValidation();
            }
            @Override public void onTextChanged(CharSequence s, int start, int before, int count) {
                checkPasswordValidation();
            }
            @Override
            public void afterTextChanged(Editable s) {
                checkPasswordValidation();
                checkFields();
            }
        });

        // password2 watcher — only updates button state
        password2.addTextChangedListener(new TextWatcher() {
            @Override public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            @Override public void onTextChanged(CharSequence s, int start, int before, int count) {}
            @Override
            public void afterTextChanged(Editable s) {
                checkFields();
            }
        });

        AndroidUtils.password(password1, iv_toggle_password1);
        AndroidUtils.password(password2, iv_toggle_password2);

        // Submit click — show alert if mismatch or invalid
        submit.setOnClickListener(view -> {
            String password_check1 = password1.getText().toString();
            String password_check2 = password2.getText().toString();

            if (!isValidPassword(password_check1)) {
                AndroidUtils.showAlert(
                        "Password does not meet the required conditions.",
                        reset_password_file.this
                );
            } else if (!password_check1.equals(password_check2)) {
                AndroidUtils.showAlert(
                        "Confirm password mismatch.",
                        reset_password_file.this
                );
            } else {
                reset_pwd();
            }
        });

        Button cancel = findViewById(R.id.Cancel);
        cancel.setOnClickListener(view -> {
            startActivity(new Intent(reset_password_file.this, LoginActivity.class));
            finish();
        });
    }

    // Conditions disappear one by one as each is met
    private void checkPasswordValidation() {
        String password = password1.getText().toString();

        if (password.isEmpty()) {
            // Show all conditions again if field is cleared
            tvErrorMsg.setVisibility(View.VISIBLE);
            tvErrorMsg.setText(
                    "• Must be 8–15 characters long\n" +
                            "• Must include uppercase and lowercase letters\n" +
                            "• Must include a number and a special character"
            );
            return;
        }

        StringBuilder errorMsg = new StringBuilder();

        if (password.length() < 8 || password.length() > 15) {
            errorMsg.append("• Must be 8–15 characters long\n");
        }
        if (!password.matches(".*[A-Z].*") || !password.matches(".*[a-z].*")) {
            errorMsg.append("• Must include uppercase and lowercase letters\n");
        }
        if (!password.matches(".*[0-9].*") || !password.matches(".*[-!@#$%&*^+=_].*")) {
            errorMsg.append("• Must include a number and a special character\n");
        }

        if (errorMsg.length() > 0) {
            tvErrorMsg.setVisibility(View.VISIBLE);
            tvErrorMsg.setText(errorMsg.toString().trim());
        } else {
            tvErrorMsg.setVisibility(View.GONE); // All conditions met ✓
        }
    }

    // Enable submit only when both fields are non-empty
    private void checkFields() {
        String p1 = password1.getText().toString().trim();
        String p2 = password2.getText().toString().trim();

        if (!p1.isEmpty() && !p2.isEmpty()) {
            submit.setEnabled(true);
            submit.setBackgroundTintList(ColorStateList.valueOf(
                    getResources().getColor(R.color.blue)));
        } else {
            submit.setEnabled(false);
            submit.setBackgroundTintList(ColorStateList.valueOf(
                    getResources().getColor(R.color.dullBlueColor)));
        }
    }

    private void reset_pwd() {
        try {
            check_url();
            Constants.PROBIZ_TYPE = "PROFESSIONAL";
            Constants.base_URL    = Constants.PROF_URL;

            JSONObject postData = new JSONObject();
            progressDialog = AndroidUtils.get_progress(reset_password_file.this);
            postData.put("old_password", Constants.OLD_PASSWORD);
            postData.put("field", "password");
            postData.put("password", Objects.requireNonNull(password1.getText().toString()));

            String urlpath = "password/" + Constants.PK + "/user/" + Constants.USER_ID + "/update";
            WebServiceHelper.callHttpWebService(
                    reset_password_file.this,
                    reset_password_file.this,
                    WebServiceHelper.RestMethodType.PUT,
                    urlpath, "UPDATE",
                    postData.toString()
            );
            Log.e("Reset Password", "Path: " + urlpath);
        } catch (Exception e) {
            if (progressDialog != null && progressDialog.isShowing())
                AndroidUtils.dismiss_dialog(progressDialog);
        }
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    @Override
    public void onClick(View view) {}

    @Override
    public void onAsyncTaskComplete(HttpResultDo httpResult) {
        if (progressDialog != null && progressDialog.isShowing())
            AndroidUtils.dismiss_dialog(progressDialog);

        if (httpResult.getResult() == WebServiceHelper.ServiceCallStatus.Success) {
            try {
                JSONObject result = new JSONObject(httpResult.getResponseContent());
                if (httpResult.getRequestType().equals("UPDATE")) {
                    if (!result.getBoolean("error")) {
                        AndroidUtils.showToast(result.getString("msg"), reset_password_file.this);

                        SharedPreferences prefs = getSharedPreferences("MyPrefs", Context.MODE_PRIVATE);
                        prefs.edit()
                                .remove("pk")
                                .remove("user_id")
                                .remove("old_password")
                                .apply();

                        Intent intent = new Intent(this, LoginActivity.class);
                        intent.putExtra("password_reset_success", true);
                        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
                        startActivity(intent);
                        finish();
                    } else {
                        AndroidUtils.showToast(
                                "Password reset failed: " + result.optString("msg"),
                                reset_password_file.this
                        );
                    }
                }
            } catch (Exception e) {
                AndroidUtils.showToast(e.getMessage(), reset_password_file.this);
            }
        } else {
            AndroidUtils.showToast(httpResult.getResponseContent(), reset_password_file.this);
        }
    }

    private boolean isValidPassword(String password) {
        if (password.length() < 8 || password.length() > 15) return false;
        if (!password.matches(".*[A-Z].*") || !password.matches(".*[a-z].*")) return false;
        if (!password.matches(".*[0-9].*") || !password.matches(".*[-!@#$%&*^+=_].*")) return false;
        return true;
    }
}

File: app/src/main/java/com/digicoffer/lauditor/LoginActivity/ViewModels/reset_password_file.kt
package com.digicoffer.lauditor.LoginActivity.ViewModels

import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.core.content.ContextCompat
import androidx.core.view.WindowCompat
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.Constants
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.Constants.check_url
import com.digicoffer.lauditor.R
import com.digicoffer.lauditor.Webservice.AsyncTaskCompleteListener
import com.digicoffer.lauditor.Webservice.HttpResultDo
import com.digicoffer.lauditor.Webservice.CommonApiHelper.WebServiceHelper
import org.json.JSONObject

class reset_password_file : AppCompatActivity(), AsyncTaskCompleteListener, View.OnClickListener {

    private var progressDialog: Dialog? = null

    // Reactive Compose States
    private var p1State by mutableStateOf("")
    private var p2State by mutableStateOf("")
    private var isLoadingState by mutableStateOf(false)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Load and validate session data from Intent and SharedPreferences
        loadSessionData()

        Log.d("RESET_PASSWORD", "onCreate Session Audit: PK=${Constants.PK}, USER_ID=${Constants.USER_ID}, OLD_PASSWORD=${Constants.OLD_PASSWORD}, TOKEN=${Constants.TOKEN?.take(10)}...")

        val composeView = androidx.compose.ui.platform.ComposeView(this).apply {
            setContent {
                com.digicoffer.lauditor.core.designsystem.theme.LauditorTheme {
                    val lengthRule = p1State.length in 8..15
                    val casingRule = p1State.matches(Regex(".*[A-Z].*")) && p1State.matches(Regex(".*[a-z].*"))
                    val numSpecialRule = p1State.matches(Regex(".*[0-9].*")) && p1State.matches(Regex(".*[-!@#$%&*^+=_].*"))

                    val reqList = listOf(
                        com.digicoffer.lauditor.core.ui.common.feedback.PasswordRequirementItem("Must be 8–15 characters long", lengthRule),
                        com.digicoffer.lauditor.core.ui.common.feedback.PasswordRequirementItem("Must include uppercase and lowercase letters", casingRule),
                        com.digicoffer.lauditor.core.ui.common.feedback.PasswordRequirementItem("Must include a number and a special character", numSpecialRule)
                    )

                    val canSubmit = p1State.isNotEmpty() && p2State.isNotEmpty()

                    com.digicoffer.lauditor.ui.auth.ResetPasswordScreen(
                        password1Value = p1State,
                        onPassword1Change = { p1State = it },
                        password2Value = p2State,
                        onPassword2Change = { p2State = it },
                        requirements = reqList,
                        isSubmitEnabled = canSubmit,
                        isLoading = isLoadingState,
                        onSubmitClick = {
                            Log.d("RESET_PASSWORD", "Submit pressed: canSubmit=$canSubmit, p1State='$p1State', p2State='$p2State'")
                            if (!isValidPassword(p1State)) {
                                Log.d("RESET_PASSWORD", "Validation Exit Point: Password rules failed for p1State='$p1State' (length=${p1State.length}, casing=$casingRule, numSpecial=$numSpecialRule)")
                                AndroidUtils.showAlert("Password does not meet the required conditions.", this@reset_password_file)
                            } else if (p1State != p2State) {
                                Log.d("RESET_PASSWORD", "Validation Exit Point: Password mismatch (p1State='$p1State', p2State='$p2State')")
                                AndroidUtils.showAlert("Confirm password mismatch.", this@reset_password_file)
                            } else {
                                Log.d("RESET_PASSWORD", "Validation Passed: Invoking resetPwd()")
                                resetPwd()
                            }
                        },
                        onCancelClick = {
                            Log.d("RESET_PASSWORD", "Cancel pressed: Navigating to LoginActivity")
                            startActivity(Intent(this@reset_password_file, LoginActivity::class.java))
                            finish()
                        }
                    )
                }
            }
        }
        setContentView(composeView)

        window.statusBarColor = ContextCompat.getColor(this, R.color.Blue_text_color)
        WindowCompat.getInsetsController(window, window.decorView)
            .isAppearanceLightStatusBars = true
        supportActionBar?.hide()
    }

    private fun loadSessionData() {
        val intentPk = intent.getStringExtra("pk")
        val intentUserId = intent.getStringExtra("user_id")

        if (!intentPk.isNullOrEmpty()) Constants.PK = intentPk
        if (!intentUserId.isNullOrEmpty()) Constants.USER_ID = intentUserId

        val prefs = getSharedPreferences("MyPrefs", Context.MODE_PRIVATE)
        if (Constants.PK.isNullOrEmpty()) Constants.PK = prefs.getString("pk", "") ?: ""
        if (Constants.USER_ID.isNullOrEmpty()) Constants.USER_ID = prefs.getString("user_id", "") ?: ""
        if (Constants.TOKEN.isNullOrEmpty()) Constants.TOKEN = prefs.getString("Token", "") ?: ""
        if (Constants.OLD_PASSWORD.isNullOrEmpty()) Constants.OLD_PASSWORD = prefs.getString("password", "") ?: ""
    }

    private fun resetPwd() {
        loadSessionData()

        if (Constants.PK.isNullOrEmpty() || Constants.USER_ID.isNullOrEmpty()) {
            Log.e("RESET_PASSWORD", "Exit Point: Session Error - PK or USER_ID is null/empty. PK='${Constants.PK}', USER_ID='${Constants.USER_ID}'")
            AndroidUtils.showAlert("Session expired. Please sign in again.", this)
            return
        }

        try {
            isLoadingState = true
            check_url()
            Constants.PROBIZ_TYPE = "PROFESSIONAL"
            Constants.base_URL = Constants.PROF_URL

            val postData = JSONObject()
            progressDialog = AndroidUtils.get_progress(this)
            postData.put("old_password", Constants.OLD_PASSWORD ?: "")
            postData.put("field", "password")
            postData.put("password", p1State)

            val urlpath = "password/${Constants.PK}/user/${Constants.USER_ID}/update"
            Log.d("RESET_PASSWORD", "API Execution: Endpoint=${Constants.base_URL}$urlpath")
            Log.d("RESET_PASSWORD", "API Execution: Request Payload=$postData")

            WebServiceHelper.callHttpWebService(
                this,
                this,
                WebServiceHelper.RestMethodType.PUT,
                urlpath,
                "UPDATE",
                postData.toString()
            )
        } catch (e: Exception) {
            isLoadingState = false
            if (progressDialog != null && progressDialog!!.isShowing) {
                AndroidUtils.dismiss_dialog(progressDialog)
            }
            e.printStackTrace()
            Log.e("RESET_PASSWORD", "Exit Point: Exception during resetPwd: ${e.message}")
        }
    }

    override fun onAsyncTaskComplete(httpResult: HttpResultDo) {
        isLoadingState = false
        if (progressDialog != null && progressDialog!!.isShowing) {
            AndroidUtils.dismiss_dialog(progressDialog)
        }

        Log.d("RESET_PASSWORD", "Raw httpResult: status=${httpResult.result}, statusCode=${httpResult.status_code}, content=${httpResult.responseContent}")

        if (httpResult.result == WebServiceHelper.ServiceCallStatus.Success) {
            try {
                val result = JSONObject(httpResult.responseContent ?: "")
                if (httpResult.requestType == "UPDATE") {
                    if (!result.getBoolean("error")) {
                        Log.d("RESET_PASSWORD", "Success Exit Point: Password updated successfully! Clearing session & navigating.")
                        AndroidUtils.showToast(result.getString("msg"), this)

                        val prefs = getSharedPreferences("MyPrefs", Context.MODE_PRIVATE)
                        prefs.edit()
                            .remove("pk")
                            .remove("user_id")
                            .remove("old_password")
                            .apply()

                        val intent = Intent(this, LoginActivity::class.java).apply {
                            putExtra("password_reset_success", true)
                            flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                        }
                        startActivity(intent)
                        finish()
                    } else {
                        Log.d("RESET_PASSWORD", "Failure Exit Point: API returned error=true, msg='${result.optString("msg")}'")
                        AndroidUtils.showToast(
                            "Password reset failed: ${result.optString("msg")}",
                            this
                        )
                    }
                }
            } catch (e: Exception) {
                Log.d("RESET_PASSWORD", "Failure Exit Point: JSON Parse Error: ${e.message}")
                AndroidUtils.showToast(e.message, this)
            }
        } else {
            Log.d("RESET_PASSWORD", "Failure Exit Point: API Network Call Failed with content='${httpResult.responseContent}'")
            AndroidUtils.showToast(httpResult.responseContent, this)
        }
    }

    override fun onClick(v: View) {
        // Not used
    }

    private fun isValidPassword(password: String): Boolean {
        if (password.length < 8 || password.length > 15) return false
        if (!password.matches(Regex(".*[A-Z].*")) || !password.matches(Regex(".*[a-z].*"))) return false
        if (!password.matches(Regex(".*[0-9].*")) || !password.matches(Regex(".*[-!@#$%&*^+=_].*"))) return false
        return true
    }
}
